neo/Consensus/Helper.cs

```diff
@@ -26,9 +27,19 @@ internal static class Helper
         public static bool CommitSent(this IConsensusContext context) => context.CommitPayloads[context.MyIndex] != null;
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static bool BlockSent(this IConsensusContext context) => context.Block != null;
+
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static bool ViewChanging(this IConsensusContext context) => context.ChangeViewPayloads[context.MyIndex]?.GetDeserializedMessage<ChangeView>().NewViewNumber > context.ViewNumber;
 
+        public static bool NotAcceptingPayloadsDueToViewChanging(this IConsensusContext context) => context.ViewChanging() && !context.MoreThanFNodesCommitted();
+
+        // A possible attack can happen if the last node to commit is malicious and either sends change view after his
+        // commit to stall nodes in a higher view, or if he refuses to send recovery messages. In addition, if a node
+        // asking change views loses network or crashes and comes back when nodes are committed in more than one higher
+        // numbered view, it is possible for the node accepting recovery and commit in any of the higher views, thus
+        // potentially splitting nodes among views and stalling the network.
+        public static bool MoreThanFNodesCommitted(this IConsensusContext context) => context.CommitPayloads.Count(p => p != null) > context.F();
```

Comment:
I think it should be `>= context.F()`. When `F` nodes are committed, if `false` is returned here, then the local node may commit, and the committed nodes may become `F+1`.

Reply:
But in that case the other nodes can join it in this view because they won’t be able to get M nodes to change view out of this view. So I think it is correct as is.

Reply:
Also, if you use >= F it causes a problem because that would allow a node to commit after sending it’s change view, possibly allowing the other nodes to move to the next view after it had committed leading to a stall.

Reply:
`commits > F` and `commits >= F` only differ when `commits == F`. I think you should not commit when `commits == F`. Otherwise it will enter the `F+1` situation. So here we should use `>= F`.

Reply:
When `commits == F`, the remaining nodes are `M` nodes which is enough to still come to consensus in the next view. Therefore this must user `> F` as the criteria for forcing coming to consensus in this view.

Reply:
It will stall potentially if you use >= F. Let me describe the scenario.

7 CN nodes
A commit in 0
B Commit in 1
C,D,E,F time-out and send change view before seeing B commit
G times out too, but then see’s B commit before he see’s other’s change view so he commits in 1
C,D,E,F see G’s change view and switch to view 2. Now it would stall.

This doesn’t happen if using `> F`

Reply:
7 CN nodes
A commit in view `0`.
B commit in view `1`.
C,D,E,F,G time-out and send `ChangeView`.
But for some reason, view changing did not succeed immediately.
Now `commits == F`, so they G commit in view `1`.
C,D,E,F see G’s change view and switch to view `2`. Now it would stall.

Reply:
If G sends change view he will not commit. In view 1.

Reply:
Which code prevented him from committing?

Reply:
All code paths that allow committing are blocked by `context.NotAcceptingPayloadsDueToViewChanging()`, since in your example only F commits are there, after G sends change view, his view will be changing, and so that flag will prevent him from committing. Specifically, Here is the code that prevents him from committing in the case of receiving `PrepareResponse`.

https://github.com/neo-project/neo/blob/b7e79810506ea1ba5eac7a837895361910d65fad/neo/Consensus/ConsensusService.cs#L416

There is similar code preventing him if `PrepareRequest` is received and if Recovery is received.

Reply:
Why do we use `NotAcceptingPayloadsDueToViewChanging()` instead of using `ViewChanging()` directly?

Reply:
Because it has to allow accepting `PrepareRequest` and `PrepareResponse` once `> F` nodes are committed, otherwise it can stall where the last nodes with view changing would never commit.

Reply:
I thought of one alternative that would allow us to use `ViewChanging()` directly; it requires causing the committed nodes to remove their commitment once an agreement has been reached to change the view. That should be safe, and will ensure liveness if `F` nodes or less have failed. It should be easy for the committed node to detect the view change by either seeing the change view messages or seeing a recovery message to a higher view. If a recovery message from a lower view is received from a committed node a recovery message could also be sent so a committed node in the lower view will change view forward.

Reply:
I'm going to make another PR alternative to this one, that maintains the liveness with`F` failed nodes.

Reply:
I'll work on it 24-30 hours from now.

Reply:
Actually I thought of how to solve it. It needs to have a new protocol message that it is like a change view commit. When `M` or those messages have been received committed nodes in lower view can uncommit.

Reply:
Actually, a recovery message that contains M `ChangeView` messages is equivalent to a Commitment to change view.

Reply:
If a recovery message is received from a higher view it will be possible to uncommit from the lower view. No new protocol messages are needed. We just need to send the recovery if we see a node committed in a lower view.

Reply:
I have an elegant solution almost complete now to maintain liveness with `F` failed nodes. Changing view needed to distinguish between a request to change view and a commitment to change view. In this way, nodes can continue to accept preparations until they commit to change view, solving liveness issues when there are up to `F` failed nodes. Also, we maintain

Reply:
Jeff, I agree. I think it is a good idea. If we have a solution that do not block `commit` forever it is much better and will respect the properties of the classical `pBFT`.
If it we proof that it works I would prefer to merge it than merging 642.
Maybe we take some more couple of days but use that. Meanwhile, @erikzhang, if you want, I think that NGD could keep testing this and put this running on CN - even if we are not going to use the other one.

@erikzhang and @igormcoelho & @shargon, let's check this possible "worst case situation" and think if this idea will cause any possibility of exposing a double header.

I will create an example of what I am seeing for us to try to create a worst case: 

1. `f_0 = |f|` honest nodes are committed in `view0`;
1. `2f+1 =|M|` nodes change view to `view1`.

Consideration:
*  `f_m = |f|` nodes from `view1` are malicious and can use their commit signatures both at `view0` and `view1`;

Implications:
* `2f = |f_0| + |f_m|` signatures can exist at `view0`;
* `f_0` nodes at `view0` can be moved to `view1` (any of the `2f+1` CN can move then to `view1`);

Security:
* We should be sure that any of the `f+1` honests that went to `view1` (2.) will have a hardware fail. In this sense, I am thinking that `PR-643` still should be merged, @jsolman. We do not want these nodes to be back because of hardware fail and give their signatures to `f_c + f_m = |M| -1` at `view0`.